libraries/protocol/custom_authorities/restriction_predicate.hxx

```diff
@@ -0,0 +1,595 @@
+   bool operator()(const A&, const B&) const { return false; }
+};
+// Equality comparison
+template<typename A, typename B, typename = void> struct predicate_eq : predicate_invalid<A, B> {};
+template<typename Field, typename Argument>
+struct predicate_eq<Field, Argument, std::enable_if_t<std::is_same<Field, Argument>::value>> {
+   // Simple comparison, same type
+   constexpr static bool valid = true;
+   constexpr bool operator()(const Field& f, const Argument& a) const { return f == a; }
+};
+template<typename Field, typename Argument>
+struct predicate_eq<Field, Argument, std::enable_if_t<is_integral<Field> && is_integral<Argument> &&
+                                                      !std::is_same<Field, Argument>::value>> {
+   // Simple comparison, integral types
+   constexpr static bool valid = true;
+   constexpr bool operator()(const Field& f, const Argument& a) const { return safenum::equal(to_num(f), to_num(a)); }
+};
+template<typename Field, typename Argument>
+struct predicate_eq<Field, Argument, std::enable_if_t<is_container<Field> && is_integral<Argument>>> {
+   // Compare container size against int
+   constexpr static bool valid = true;
+   bool operator()(const Field& f, const Argument& a) const { return safenum::equal(f.size(), to_num(a)); }
+};
+template<typename Field, typename Argument>
+struct predicate_eq<fc::optional<Field>, Argument, std::enable_if_t<comparable_types<Field, Argument>>>
+   : predicate_eq<Field, Argument> {
+   // Compare optional value against comparable type
+   using base = predicate_eq<Field, Argument>;
+   bool operator()(const fc::optional<Field>& f, const Argument& a) const {
+      if (!f.valid()) return predicate_result::Rejection(predicate_result::null_optional);
+      return (*this)(*f, a);
+   }
+};
+template<typename Field>
+struct predicate_eq<fc::optional<Field>, void_t, void> {
+   // Compare optional value against void_t (checks that optional is null)
+   constexpr static bool valid = true;
+   bool operator()(const fc::optional<Field>& f, const void_t&) const { return !f.valid(); }
+};
+// Not-equal is just an equality comparison wrapped in a negator
+template<typename Field, typename Argument> struct predicate_ne : predicate_eq<Field, Argument> {
+   using equal = predicate_eq<Field, Argument>;
+   bool operator()(const Field& f, const Argument& a) const { return !equal::operator()(f, a); }
+};
+
+// Shared implementation for all inequality comparisons
+template<typename A, typename B, typename = void> struct predicate_compare : predicate_invalid<A, B> {};
+template<typename Field, typename Argument>
+struct predicate_compare<Field, Argument, std::enable_if_t<std::is_same<Field, Argument>::value>> {
+   // Simple comparison, same types
+   constexpr static bool valid = true;
+   constexpr int8_t operator()(const Field& f, const Argument& a) const {
+      return f<a? -1 : (f>a? 1 : 0);
```

Comment:
It's possible that `operator >` is not implemented in some data types. I guess it's better to use `a<f` here.

Reply:
Done: 71de24b